include/exec/at_coroutine_exit.hpp

```diff
@@ -226,4 +227,188 @@ namespace exec {
+      }
+
+      bool await_ready() const noexcept {
+        return false;
+      }
+
+      template <class _Promise>
+        requires __promise_base_test_t<_Promise>::value
+      bool await_suspend(__coro::coroutine_handle<_Promise> __parent) noexcept {
+        __coro_.promise().__parent_ = __parent;
+        __coro_.promise().__get_index_callback_ = [](void* __parent) noexcept -> int {
+          auto p = __coro::coroutine_handle<_Promise>::from_address(__parent).promise();
+          return p.__data_.index();
+        };
+        __coro_.promise().set_continuation(__parent.promise().continuation());
+        __parent.promise().set_continuation(__coro_);
+        return false;
+      }
+
+      std::tuple<_Ts&...> await_resume() noexcept {
+        return std::exchange(__coro_, {}).promise().__args_;
+      }
+
+     private:
+      struct __final_awaitable {
+        static std::false_type await_ready() noexcept {
+          return {};
+        }
+
+        static __coro::coroutine_handle<>
+          await_suspend(__coro::coroutine_handle<__promise> __h) noexcept {
+          __promise& __p = __h.promise();
+          auto __coro = __p.__is_unhandled_stopped_
+                        ? __p.continuation().unhandled_stopped()
+                        : __p.continuation().handle();
+          __h.destroy();
+          return __coro;
+        }
+
+        void await_resume() const noexcept {
+        }
+      };
+
+      struct __promise : with_awaitable_senders<__promise> {
+        template <class _Action>
+        explicit __promise(_Action&&, _Ts&... __ts) noexcept
+          : __args_{__ts...} {
+        }
+
+        __coro::suspend_always initial_suspend() noexcept {
+          return {};
+        }
+
+        __final_awaitable final_suspend() noexcept {
+          return {};
+        }
+
+        void return_void() noexcept {
+        }
+
+        [[noreturn]] void unhandled_exception() noexcept {
+          std::terminate();
+        }
+
+        __coro::coroutine_handle<__promise> unhandled_stopped() noexcept {
+          __is_unhandled_stopped_ = true;
+          return __coro::coroutine_handle<__promise>::from_promise(*this);
+        }
+
+        __task get_return_object() noexcept {
+          return __task(__coro::coroutine_handle<__promise>::from_promise(*this));
+        }
+
+        template <class _Awaitable>
+        decltype(auto) await_transform(_Awaitable&& __awaitable) noexcept {
+          STDEXEC_ASSERT(__parent_ && __get_index_callback_);
+          const int __index = __get_index_callback_(__parent_.address());
+          using __any_sender_t = typename any_receiver_ref<
+            completion_signatures<set_value_t(), set_error_t(std::exception_ptr)>>::any_sender<>;
+          __any_sender_t __sender = just();
+          if (__from_int(__index) == _OnCompletion) {
+            __sender = __at_coro_exit::__die_on_stop((_Awaitable&&) __awaitable);
+          }
```

Comment:
This is a good place to use a variant sender once we have one. Instead, I would just give the `__die_on_stopped` sender adaptor a runtime `bool` to decide whether a stopped signal should either be forwarded or else `terminate()`.

Reply:
briliant, gimme a second